build/configure/app-configure-step.js

```diff
@@ -0,0 +1,138 @@
+'use strict';
+
+var utils = require('utils');
+
+var DependencyGraph = require('./dependency-graph');
+var BuildConfig = require('./build-config');
+
+// Default building steps before app's default building steps.
+// TODO: We should have a module to map these steps with different app's own
+//       config data, so that app could easily to add/remove/reorder steps for
+//       different situations.
+const POST_APP_BUILD_STEPS = [
+  'media-resolution',
+  'post-manifest',
+  'multilocale',
+  'copy-build-stage-data',
+  'webapp-optimize',
+  'webapp-zip'
+];
+
+/**
+ * AppConfigureStep would generate all the data for build back-end, ideally it
+ * should only manage the steps relating to app itself, like app's own
+ * buildscript and POST_APP_BUILD_STEPS.
+ *
+ * @param {object} options - BUILD_CONFIG from main Makefile.
+ * @param {string} appDir - the app's source dir path.
+ * @constructor
+ */
+var AppConfigureStep = function(options, appDir) {
+  this.appDir = appDir;
+  this.webapp = utils.getWebapp(appDir, options);
+  this.name = this.webapp.sourceDirectoryName;
+  this.options = options;
+  this.buildConfig = new BuildConfig();
+};
+
+AppConfigureStep.prototype = {
+  /**
+   * Configure start from here.
+   */
+  start: function() {
+    var options = this.options;
+    this._preSteps = [];
+    this.appDirFile = utils.getFile(this.appDir);
+    this.xpcshell = utils.getEnv('XPCSHELLSDK');
+
+    // we'll execute our current modules with stringified options.
+    this.appOptions = utils.cloneJSON(options);
+
+    this.appOptions.APP_DIR = this.appDir,
+    this.appOptions.STAGE_APP_DIR = utils.joinPath(this.options.STAGE_DIR,
+      this.appDirFile.leafName);
+    this.appOptions.webapp = this.webapp;
+    // We may encounter error while stringify manifest , put it to command line
+    // and parse it back to json object, so we simply delete these since the
+    // following steps don't need these attributes.
+    delete this.appOptions.webapp.manifest;
+    delete this.appOptions.webapp.metaData;
+
+    utils.ensureFolderExists(utils.getFile(this.appOptions.STAGE_APP_DIR));
+
+    this.buildConfig.addConfig('APP_OPTIONS', this.appOptions);
+
+    // We have to rewrite below two ENV since they're using abspath and CURDIR.
+    this.buildConfig.addConfig('GAIA_DIR',
+      options.GAIA_DIR);
+    this.buildConfig.addConfig('XPCSHELLSDK',
+      options.XPCSHELLSDK);
+
+    // Create app's own makefile and store in STAGE_APP_DIR/app.mk.
+    this.appMkPath = utils.joinPath(this.appOptions.STAGE_APP_DIR, 'Makefile');
+    this.mainMake = new DependencyGraph(this.appMkPath);
+
+    this.genAppConfig();
+
+    this.mainMake.genBackend(this.buildConfig.getOutput('makefile'));
+  },
+
+  get execMakeCommand() {
```

Comment:
I'm afraid that why we named a function return a command with `exec-` prefix. If it actually does execution, we don't need to make it as a getter. If it doesn't, we don't need the prefix.